core/blockchain.go

```diff
@@ -1252,33 +1253,14 @@ func (bc *BlockChain) InsertReceiptChain(blockChain types.Blocks, receiptChain [
 
 var lastWrite uint64
 
-// writeBlockWithoutState writes only the block and its metadata to the database,
-// but does not write any state. This is used to construct competing side forks
-// up to the point where they exceed the canonical total difficulty.
-func (bc *BlockChain) writeBlockWithoutState(block *types.Block, td *big.Int) (err error) {
-	bc.wg.Add(1)
-	defer bc.wg.Done()
-
-	batch := bc.db.NewBatch()
-	rawdb.WriteTd(batch, block.Hash(), block.NumberU64(), td)
-	rawdb.WriteBlock(batch, block)
-	if err := batch.Write(); err != nil {
-		log.Crit("Failed to write block into disk", "err", err)
-	}
-	return nil
-}
-
-// writeKnownBlock updates the head block flag with a known block
-// and introduces chain reorg if necessary.
+// writeKnownBlock updates the head block flag with a known block.
 func (bc *BlockChain) writeKnownBlock(block *types.Block) error {
 	bc.wg.Add(1)
 	defer bc.wg.Done()
 
 	current := bc.CurrentBlock()
 	if block.ParentHash() != current.Hash() {
-		if err := bc.reorg(current, block); err != nil {
-			return err
-		}
+		return errParentNotCanonical
```

Comment:
As per my previous comment about error values, I think something like this would make a much more useful error.

```suggestion
		return fmt.Errorf("failed to write block %s, expected parent hash %s, current head %s",
			block.ShortString(),
			block.ParentHash().ShortString(),
			current.ShortString())
```
Supporting functions.
```go
// ShortString returns a hex representation of the first 8 bytes of the hash. Its intent is to aid debugging by providing enough of the hash that it can be easily reconciled with logs containing the full hash but is also small enough to keep the logs human readable.
func (h Hash) ShortString() string { return hexutil.Encode(h[:8]) }

// ShortString returns the block number and short hash separated b a colon.
func (b *Block) ShortString() string { return fmt.Sprintf("%d:%s", b.NumberU64(), b.Hash().ShortString()) }
```